src/gt4py/next/iterator/transforms/prune_casts.py

```diff
@@ -21,24 +35,27 @@ class PruneCasts(PreserveLocationVisitor, NodeTranslator):
     """
 
     def visit_FunCall(self, node: ir.FunCall) -> ir.Node:
-        node = self.generic_visit(node)
-
-        if not cpm.is_call_to(node, "cast_"):
-            return node
-
-        value, type_constructor = node.args
-
-        assert (
-            value.type
-            and isinstance(type_constructor, ir.SymRef)
-            and (type_constructor.id in ir.TYPEBUILTINS)
-        )
-        dtype = ts.ScalarType(kind=getattr(ts.ScalarKind, type_constructor.id.upper()))
-
-        if value.type == dtype:
-            return value
-
-        return node
+        if cpm.is_op_as_fieldop(node, "cast_"):
+            # Prune trivial `as_fieldop` cast expressions with the form:
+            # as_fieldop(λ(__val) = cast_(deref(__val), float64)))(a)
+            # where 'a' is already a field with data type float64, in this example.
+            lambda_node = node.fun.args[0]  # type: ignore[attr-defined]
+            assert isinstance(lambda_node, ir.Lambda)
+            cast_expr = lambda_node.expr
+            assert isinstance(cast_expr, ir.FunCall)
+            cast_value = cast_expr.args[0]
+            if (
+                _is_cast_redundant(cast_expr)
+                and cpm.is_call_to(cast_value, "deref")
+                and cpm.is_ref_to(cast_value.args[0], lambda_node.params[0].id)
+            ):
+                return self.visit(node.args[0])
```

Comment:
```suggestion
        if cpm.is_identity_as_fieldop(node):
            return node.args[0]
```

```python
def is_identity_as_fieldop(node: itir.Expr):
  if not is_applied_as_fieldop(node):
    return False
  stencil = node.fun.args[0]
  if is_ref_to(stencil, "deref"):
    return True
  if isinstance(stencil, itir.Lambda) and len(stencil.params) == 1 and stencil == im.lambda(stencil.params[0])(im.deref(stencil.params[0].id)):
    return True
  return False
```

Just a suggestion to avoid code duplication. The if clause here does the same `_is_cast_redundant(node)` check as below while the suggested version simply removes `as_fieldop(deref)(a)`.